digiroad2-api-viite/src/main/scala/fi/liikennevirasto/digiroad2/ViiteApi.scala

```diff
@@ -89,32 +95,79 @@ class ViiteApi(val roadLinkService: RoadLinkService, val vVHClient: VVHClient,
+        val roadNumber = data.get("roadNumber").get.asInstanceOf[Double].toLong
+        val roadPartNumber = data.get("roadPartNumber").get.asInstanceOf[Double].toLong
+        val trackCode = data.get("trackCode").get.asInstanceOf[Double].toLong
+
+        roadAddressService.getFloatingAdjacent(chainLinks, linkId, roadNumber, roadPartNumber, trackCode).map(roadAddressLinkToApi)
     }
 
   get("/roadlinks/multiSourceAdjacents") {
-      val roadData = JSON.parseFull(params.get("roadData").get).get.asInstanceOf[Seq[Map[String,Double]]]
+      val roadData = JSON.parseFull(params.get("roadData").get).get.asInstanceOf[Seq[Map[String,Any]]]
       if (roadData.isEmpty){
         Set.empty
       } else {
         val adjacents:Seq[RoadAddressLink] = {
           roadData.flatMap(rd => {
-            roadAddressService.getFloatingAdjacent(rd.get("linkId").get.toLong,
-              rd.get("roadNumber").get.toLong, rd.get("roadPartNumber").get.toLong, rd.get("trackCode").get.toLong)
+            val chainLinks = rd.get("selectedLinks").get.asInstanceOf[Seq[Double]].map(rl => {
+              rl.toLong
+            }).toSet[Long]
+            val linkId = rd.get("linkId").get.asInstanceOf[Double].toLong
+            val roadNumber = rd.get("roadNumber").get.asInstanceOf[Double].toLong
+            val roadPartNumber = rd.get("roadPartNumber").get.asInstanceOf[Double].toLong
+            val trackCode = rd.get("trackCode").get.asInstanceOf[Double].toLong
+            roadAddressService.getFloatingAdjacent(chainLinks, linkId,
+              roadNumber, roadPartNumber, trackCode, false)
           })
         }
-        val linkIds: Seq[Long] = roadData.map(rd => rd.get("linkId").get.toLong)
+        val linkIds: Seq[Long] = roadData.map(rd => rd.get("linkId").get.asInstanceOf[Double].toLong)
         val result = adjacents.filter(adj => {
           !linkIds.contains(adj.linkId)
-        })
+        }).distinct
         result.map(roadAddressLinkToApi)
       }
   }
 
+  get("/roadlinks/transferRoadLink") {
+    val (sources, targets) = roadlinksData()
+    val user = userProvider.getCurrentUser()
+    val result = roadAddressService.getRoadAddressAfterCalculation(sources, targets, user)
+    result.map(roadAddressLinkToApi)
+  }
+
+  put("/roadlinks/roadaddress") {
+    try{
+      val test = parsedBody.extract[newAddressDataExtractor]
+      val roadAddressData = test.roadAddress
+      val sourceIds = test.sourceIds
+      val targetIds = test.targetIds
+      val roadAddresses = roadAddressData.map{ ra =>
+        RoadAddress(ra.id, ra.roadNumber, ra.roadPartNumber, Track.apply(ra.trackCode), Discontinuity.apply(ra.discontinuity), ra.startAddressM, ra.endAddressM,
+        Some(DateTime.now()), None, Option(ra.modifiedBy), ra.linkId, ra.startMValue, ra.endMValue, SideCode.apply(ra.sideCode), ra.calibrationPoints, false, ra.points)
+      }
+
+      roadAddressService.transferFloatingToGap(sourceIds, targetIds, roadAddresses)
+    } catch {
```

Comment:
Is this try catch mandatory, as in, if this method could throw an exception it should be fixed and patched.